### DISCUSSION TITLE
TeamTalk v5.14 with encrypted server verification

### DISCUSSION THREAD

**Discussion by bear101** [BearWare/TeamTalk5#1920] [Announcements]
The TeamTalk clients have for a long time been able to use encrypted connections. Having an encrypted connection is, however, not enough to guarantee security because you don't know to whom you've established the connection. A "man-in-the-middle" would be able to fake a TeamTalk server and steal credentials by hijacking your initial connection attempt.

In the latest TeamTalk v5.14 it is now possible to verify that the encrypted server is authentic by having the client check the server's certificate. When connecting to an encrypted TeamTalk Pro Server you can now check "Verify Server Certificate" in setup of the encrypted connection.

Note that when generating .tt file you'll be able to include the certificate of your TeamTalk Pro Server, so users don't have to go though the hassle of setting up the certificate file. Also when you publish your TeamTalk Pro Server as an unofficial server the certificate can be published initially without user intervention.

I've written a long guide [here](https://github.com/BearWare/TeamTalk5/blob/master/Setup/Server/Linux/README_professional) that explains how to set up encrypted server verification. This README file is also included with TeamTalk Pro Server.

Also in the client's [manual](https://bearware.dk/beta/docs/md_userinterface.html#setupencryptiondlg) I've specified the different requirements for setting up an encrypted verifiable connection.

Let me know if there's something in the manual that needs extra explanation.

Good luck!
-- Bjørn

**Comment by bear101** [comment]
Btw, the new beta can be downloaded [here](https://bearware.dk/beta/) (in case you haven't see the new beta-release in TeamTalk client).

**Comment by beqabeqa473** [comment]
Hi Bjoern.

I've decided to move to encryypted server and today i've set up test server.

I did everything what was written in the readme, but for some reason certificate cannot be verified and connection doesn't happen.
Server is installed on ubuntu 20.04, client side is a qt client on windows 10

**Reply by bear101** [reply]
What error do you get in the Qt client when connecting? In status messages you should see an encryption (OpenSSL) error.

**Comment by beqabeqa473** [comment]
2023-08-30 22:40:38  * Secure connection failed due to error
0x1416f086: error:1416F086:SSL
routines:tls_process_server_certificate:certificate verify failed.

I generated certificates as explained in readme, however i didn't make
a challenge password when creating a request.
I downloaded pro server from official sources, didn't try to use my own build.
openssl on ubuntu 20.04 is 1.1.1f

On 8/30/23, Bjørn Damstedt Rasmussen ***@***.***> wrote:
> What error do you get in the Qt client when connecting? In status messages
> you should see an encryption (OpenSSL) error.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/BearWare/TeamTalk5/discussions/1920#discussioncomment-6867346
> You are receiving this because you commented.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Reply by bear101** [reply]
What output do you get if run this command:
`openssl verify -CAfile ca.cer ttservercert.pem`
Where `ca.cer` is your CA certificate and `ttservercert.pem` is your server's certificate.

**Reply by bear101** [reply]
You should get this output:
`bdr@macmini Server> openssl verify -CAfile ca.cer ../Library/TeamTalkLib/test/ttservercert.pem 
../Library/TeamTalkLib/test/ttservercert.pem: OK`

**Comment by beqabeqa473** [comment]
Thanks for the advice.

I didn't read docs of openssl.

maybe it would be better to make note of that in installation procedure.

Whatever method you use to generate the certificate and key files, the
Common Name value used for the server and client certificates/keys
must each differ from the Common Name value used for the CA
certificate. Otherwise, the certificate and key files will not work
for servers compiled using OpenSSL.

On 8/31/23, Bjørn Damstedt Rasmussen ***@***.***> wrote:
> You should get this output:
> ***@***.*** Server> openssl verify -CAfile ca.cer
> ../Library/TeamTalkLib/test/ttservercert.pem
> ../Library/TeamTalkLib/test/ttservercert.pem: OK`
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/BearWare/TeamTalk5/discussions/1920#discussioncomment-6868760
> You are receiving this because you commented.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by beqabeqa473** [comment]
Can i allow clients to only connect if server verification succeeds

On 8/31/23, Beqa Gozalishvili ***@***.***> wrote:
> Thanks for the advice.
>
> I didn't read docs of openssl.
>
> maybe it would be better to make note of that in installation procedure.
>
> Whatever method you use to generate the certificate and key files, the
> Common Name value used for the server and client certificates/keys
> must each differ from the Common Name value used for the CA
> certificate. Otherwise, the certificate and key files will not work
> for servers compiled using OpenSSL.
>
> On 8/31/23, Bjørn Damstedt Rasmussen ***@***.***> wrote:
>> You should get this output:
>> ***@***.*** Server> openssl verify -CAfile ca.cer
>> ../Library/TeamTalkLib/test/ttservercert.pem
>> ../Library/TeamTalkLib/test/ttservercert.pem: OK`
>>
>> --
>> Reply to this email directly or view it on GitHub:
>> https://github.com/BearWare/TeamTalk5/discussions/1920#discussioncomment-6868760
>> You are receiving this because you commented.
>>
>> Message ID:
>> ***@***.***>
>
>
> --
> with best regards Beqa Gozalishvili
> Tell: +995593454005
> Email: ***@***.***
> Web: https://gozaltech.org
> Skype: beqabeqa473
> Telegram: https://t.me/gozaltech
> facebook: https://facebook.com/gozaltech
> twitter: https://twitter.com/beqabeqa473
> Instagram: https://instagram.com/beqa.gozalishvili
>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Reply by bear101** [reply]
You cannot force it from the server side since it's a client setting. You can, however, automatically enable it in a .tt file or when you publish your server.

**Comment by beqabeqa473** [comment]
Yes, i already understand that.

i"ll automate certificate creation for client verification and this will work.

On 9/3/23, Bjørn Damstedt Rasmussen ***@***.***> wrote:
> You cannot force it from the server side since it's a client setting. You
> can, however, automatically enable it in a .tt file or when you publish your
> server.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/BearWare/TeamTalk5/discussions/1920#discussioncomment-6895496
> You are receiving this because you commented.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by computertech1012** [comment]
Hi,
It seams that the Stats bot is down, and Bearware web logins are not working. Whenever I try to connect to the encrypted server, it says connected, but nothing happens.
Also, for the TeamTalk server with client varification I think this should be an official server rather than an unofficial server. This server is also hosted by bearware.dk and I think it should be a third server to use.
Thanks for reading!

**Reply by bear101** [reply]
Hm, it looks like a Linode issue. I can't even ping www.bearware.dk from my Linode server

**Comment by beqabeqa473** [comment]
are you sure dns are ok on vm?
Try to ser 8.8.8.8 and 8.8.4.4 in resolve.conf and try to restart
systemd-resolved.

I had a problem like that several days ago.

On 10/6/23, Bjørn Damstedt Rasmussen ***@***.***> wrote:
> Hm, it looks like a Linode issue. I can't even ping www.bearware.dk from my
> Linode server
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/BearWare/TeamTalk5/discussions/1920#discussioncomment-7212695
> You are receiving this because you commented.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Reply by bear101** [reply]
The IP can be resolved. It's the routing that appears to be wrong

**Comment by beqabeqa473** [comment]
aah, ok, understood.

On 10/7/23, Bjørn Damstedt Rasmussen ***@***.***> wrote:
> The IP can be resolved. It's the routing that appears to be wrong
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/BearWare/TeamTalk5/discussions/1920#discussioncomment-7213493
> You are receiving this because you commented.
>
> Message ID:
> ***@***.***>


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: ***@***.***
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili
